### PR TITLE
OCPBUGS-18131: Block upgrade to 4.13 via  admin ack

### PR DESCRIPTION
**Pull request by gnufied**
Fixes https://issues.redhat.com/browse/OCPBUGS-18131

### CODE REVIEW

File: pkg/operator/vspherecontroller/checks/check_error.go [hunk 1]
```diff
@@ -162,11 +189,62 @@ func CheckClusterStatus(result ClusterCheckResult, apiDependencies KubeAPIInterf
...
+
+// returns false is migration is enabled in the cluster.
+// returns true if migration is not enabled in the cluster and cluster is using
+// in-tree vSphere volumes.
+func checkForIntreePluginUse(result ClusterCheckResult, apiDependencies KubeAPIInterface) (ClusterCheckStatus, ClusterCheckResult) {
+	allGoodCheckResult := MakeClusterCheckResultPass()
+
+	pvs, err := apiDependencies.ListPersistentVolumes()
+	if err != nil {
+		reason := fmt.Errorf("vsphere csi driver installed failed with %s, unable to list pvs: %v", result.Reason, err)
```
Comment (jsafrane) [inline pkg/operator/vspherecontroller/checks/check_error.go:209]: This error is IMO not related to CSI driver installation
(this applies also to the original PR, so I don't mind if we leave it as it is...)

**Review by jsafrane — COMMENTED** [review]
/lgtm

**Comment by duanwei33** [comment]
Pre-merge verified with PR: openshift/vmware-vsphere-csi-driver-operator#171 and openshift/cluster-storage-operator#402.

Create cluster in vsphere 8, while ESXi version is 8.0.1.0-22088125(doesn't match the requirement) and vsphere version is 8.0.1.0-22088981 (matches the requirement) 
1. **oc adm upgrade** shows the CSI migration warning as below which is expected, not starting with a new line however:
```
$ oc adm upgrade
Cluster version is 4.12.0-0.test-2023-09-28-022513-ci-ln-58nnh9k-latest

Upgradeable=False

  Reason: AdminAckRequired
  Message: Kubernetes 1.26 and therefore OpenShift 4.13 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6958394 for details and instructions. vSphere CSI migration will be enabled in Openshift-4.14. Your cluster appears to be using in-tree vSphere volumes and is on a vSphere version that has CSI migration related bugs. See - https://access.redhat.com/node/7011683 for more information, before upgrading to 4.14.
```
2. Check the driver operator log which is expected:
```
I0928 07:16:18.276872       1 vmware.go:273] Found existing profile with same name: openshift-storage-policy-wduan-0928b-vs-k6pft
I0928 07:16:19.178787       1 check_nodes.go:174] checking for patched version of ESXi for CSI migration: 8.0.1.0-22088125
I0928 07:16:19.178913       1 check_nodes.go:174] checking for patched version of ESXi for CSI migration: 8.0.1.0-22088125
I0928 07:16:19.186219       1 check_nodes.go:174] checking for patched version of ESXi for CSI migration: 8.0.1.0-22088125
I0928 07:16:19.186276       1 check_patched_vcenter.go:29] checking for patched version of vSphere for CSI migration: 8.0.1.0-22088981
I0928 07:16:19.186289       1 vsphere_environment_checker.go:90] Overall check result: UpgradeRequiresAdminAck: host host-1013 is on ESXI version 8.0.1.0-22088125, minimum required for CSI migration: > 8.0.1-22088125
I0928 07:16:19.186296       1 vspherecontroller.go:261] Scheduled the next check in 1h0m0s
I0928 07:16:19.186310       1 check_error.go:217] found vSphere in-tree persistent volumes in the cluster
I0928 07:16:19.186316       1 vspherecontroller.go:361] check status is: pass
I0928 07:16:19.186333       1 vspherecontroller.go:516] Updating admin-gates to require admin-ack for vSphere CSI migration
W0928 07:16:19.199366       1 vspherecontroller.go:496] host host-1013 is on ESXI version 8.0.1.0-22088125, minimum required for CSI migration: > 8.0.1-22088125
I0928 07:16:19.199788       1 event.go:285] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"openshift-cluster-csi-drivers", Name:"vmware-vsphere-csi-driver-operator", UID:"4a86e840-d6e9-4af9-b1a5-8eac2ad524ea", APIVersion:"apps/v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'ConfigMapUpdated' Updated ConfigMap/admin-gates -n openshift-config-managed:
```

Check admin-gates, we use   **ack-4.12-kube-126-vsphere-migration-in-4.14** as it has impact when upgrading to 4.14 which is expected:
```
$ oc get ConfigMap/admin-gates -n openshift-config-managed -o yaml
apiVersion: v1
data:
  ......
  ack-4.12-kube-126-vsphere-migration-in-4.14: vSphere CSI migration will be enabled
    in Openshift-4.14. Your cluster appears to be using in-tree vSphere volumes and
    is on a vSphere version that has CSI migration related bugs. See - https://access.redhat.com/node/7011683
    for more information, before upgrading to 4.14.
```

**Comment by duanwei33** [comment]
@gnufied Just one strange observation, the message "Scheduled the next check in 1h0m0s" which stands for the end actually came before all check finished:

```
I0928 07:16:18.276872       1 vmware.go:273] Found existing profile with same name: openshift-storage-policy-wduan-0928b-vs-k6pft
I0928 07:16:19.178787       1 check_nodes.go:174] checking for patched version of ESXi for CSI migration: 8.0.1.0-22088125
I0928 07:16:19.178913       1 check_nodes.go:174] checking for patched version of ESXi for CSI migration: 8.0.1.0-22088125
I0928 07:16:19.186219       1 check_nodes.go:174] checking for patched version of ESXi for CSI migration: 8.0.1.0-22088125
I0928 07:16:19.186276       1 check_patched_vcenter.go:29] checking for patched version of vSphere for CSI migration: 8.0.1.0-22088981
I0928 07:16:19.186289       1 vsphere_environment_checker.go:90] Overall check result: UpgradeRequiresAdminAck: host host-1013 is on ESXI version 8.0.1.0-22088125, minimum required for CSI migration: > 8.0.1-22088125
I0928 07:16:19.186296       1 vspherecontroller.go:261] Scheduled the next check in 1h0m0s
I0928 07:16:19.186310       1 check_error.go:217] found vSphere in-tree persistent volumes in the cluster
I0928 07:16:19.186316       1 vspherecontroller.go:361] check status is: pass
I0928 07:16:19.186333       1 vspherecontroller.go:516] Updating admin-gates to require admin-ack for vSphere CSI migration
W0928 07:16:19.199366       1 vspherecontroller.go:496] host host-1013 is on ESXI version 8.0.1.0-22088125, minimum required for CSI migration: > 8.0.1-22088125
I0928 07:16:19.199788       1 event.go:285] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"openshift-cluster-csi-drivers", Name:"vmware-vsphere-csi-driver-operator", UID:"4a86e840-d6e9-4af9-b1a5-8eac2ad524ea", APIVersion:"apps/v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'ConfigMapUpdated' Updated ConfigMap/admin-gates -n openshift-config-managed:
```

**Comment by gnufied** [comment]
@duanwei33 that should be okay. So we run all the expensive checks and when they are finished we run another check which takes result of previously finished checks and applies further logic. For example - if expensive checks resulted in determining an older version of vSphere then the next check verifies if cluster has intree PVs.

**Pull request merged**